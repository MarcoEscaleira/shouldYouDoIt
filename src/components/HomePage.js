import React, { Component, Fragment } from 'react';
import Form from './Form';
import Answer from './Answer';
import ErrorModal from './ErrorModal';

const DEFAULT_STATE = {
  print: false,
  data: {},
  oldTextInput: '',
  textInput: '',
  typeError: {
    showModal: false,
    text: ''
  }
};

class Home extends Component {
  state = {
    ...DEFAULT_STATE
  };
  
  handleFormSubmit = (e) => {
    e.preventDefault();

    this.handleData();
    
    const inputText = e.target.toDo.value.trim();

    // Check if input is empty
    if (!inputText) {
      this.handleData();
      this.handleTypeError("Please enter what you want to do");
      return;
    }

    // Check if input is same as before
    if (this.state.oldTextInput === inputText) {
      this.handleData();
      this.handleTypeError("Please do another thing!");
      return;
    }

    //Loading animation 
    this.setState(() => ({ print: true, data: {
      "msg": "loading",
      "img": "./gifs/Blocks-1s-200px.gif"
    } }));
    
    const request = async () => {
      const response = await fetch(`https://shouldyoudoit.herokuapp.com/?search=${inputText}`, {
        headers: {
          "Content-Type": "application/json",
        }
      });
      const data = await response.json();
      this.setState(() => ({ print: true, data, oldTextInput: inputText }));
    };
    request();
  };

  handleState = () => this.setState(() => ({ ...DEFAULT_STATE }));
  
  handleData = () => this.setState(() => ({ 
    print: false,
    data: {}
  }));

  handleTypeError = (text = '') => this.setState(() => ({
    typeError: {
      showModal: true,
      text
    }
  }));

  closeTypeError = () => this.setState(() => ({
    typeError: {
      showModal: false,
      text: ''
    }
  }));

  handleTextInput = (e) => {
    const textInput = e.target.value;
    if(!textInput || textInput.match(/\w+/)) {
      this.setState(() => ({ textInput }));
    }
  };

  render() {
    return (
      <Fragment>
        <div className="container">
          <Form 
            handleFormSubmit={this.handleFormSubmit} 
            handleTextInput={this.handleTextInput} 
            text={this.state.textInput}
          />
          { this.state.print &&
            <Answer 
              handleData={this.handleState} 
              data={this.state.data} 
            />
          }
        </div>
        <ErrorModal 
          show={this.state.typeError.showModal}
          text={this.state.typeError.text}
          closeCallback= {this.closeTypeError}
        />
      </Fragment>
    );
  }
}

export default Home;
