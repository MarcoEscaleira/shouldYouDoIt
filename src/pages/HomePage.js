import React from 'react';
import { useSelector } from 'react-redux';

import Form from '../components/Form';
import Answer from '../components/Answer';

const Home = () => {
  const print = useSelector(state => state.print);
  return (
    <div className="container">
      <Form />
      {print && <Answer />}
    </div>
  );
};

export default Home;
