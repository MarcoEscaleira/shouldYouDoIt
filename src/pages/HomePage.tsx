import React from 'react';
import { useSelector } from 'react-redux';

import Form from '../components/Form';
import Answer from '../components/Answer';
import { DecisionsState } from '~store/decisions';

const Home: React.FC = () => {
  const print = useSelector<DecisionsState, boolean>(state => state.print);

  return (
    <section className="container">
      <Form />
      {print && <Answer />}
    </section>
  );
};

export default Home;
