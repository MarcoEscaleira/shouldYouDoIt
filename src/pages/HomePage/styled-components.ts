import styled from "styled-components";

export const Container = styled.section`
  width: 100%;
  height: 100%;
  flex-grow: 1;
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
`;
