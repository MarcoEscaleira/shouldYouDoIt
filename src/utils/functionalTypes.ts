export type NullaryFn<R> = () => R;
export type UnaryFn<A, R> = (a: A) => R;
export type BinaryFn<A, B, R> = (a: A, b: B) => R;
export type TernaryFn<A, B, C, R> = (a: A, b: B, c: C) => R;
export type QuadFn<A, B, C, D, R> = (a: A, b: B, c: C, d: D) => R;
export type QuinFn<A, B, C, D, F, R> = (a: A, b: B, c: C, d: D, f: F) => R;
export type SixFn<A, B, C, D, F, H, R> = (
  a: A,
  b: B,
  c: C,
  d: D,
  f: F,
  h: H
) => R;
