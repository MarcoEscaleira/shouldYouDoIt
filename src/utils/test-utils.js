import React from 'react';
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import { MemoryRouter as Router } from 'react-router-dom';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import store from '../store';

export const renderWithRedux = (ui, initialState) => {
  return {
    ...render(
      <Provider store={store(initialState)}>
        <Router>{ui}</Router>
      </Provider>
    )
  };
};

export const renderWithRouter = ui => {
  return {
    ...render(<Router>{ui}</Router>)
  };
};

const middlewares = [thunk];
export const mockStore = configureStore(middlewares);
