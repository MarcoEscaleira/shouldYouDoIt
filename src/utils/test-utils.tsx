import React from "react";
import { render } from "@testing-library/react";
import { Provider } from "react-redux";
import { MemoryRouter as Router } from "react-router-dom";
import configureStore from "redux-mock-store";
import thunk from "redux-thunk";

import store, { initialStoreState } from "../store";

export const renderWithRedux = (ui: JSX.Element, initialState?: {}) => {
  const state = {
    ...initialStoreState,
    ...initialState
  };
  return {
    ...render(
      <Provider store={store(state)}>
        <Router>{ui}</Router>
      </Provider>
    )
  };
};

export const renderWithRouter = (ui: JSX.Element) => ({
  ...render(<Router>{ui}</Router>)
});

const middlewares = [thunk];
export const mockStore = configureStore(middlewares);
