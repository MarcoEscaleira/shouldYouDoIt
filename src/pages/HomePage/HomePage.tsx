import React, { Fragment } from "react";
import { connect } from "react-redux";
import { Container } from "./styled-components";

import Form from "~/components/Form/Form";
import Answer from "~/components/Answer/Answer";
import ErrorModal from "~/components/ErrorModal/ErrorModal";
import { StoreState } from "~/store/";
import { getPrint } from "~/store/decisions/selectors";
import { getErrorModal } from "~/store/errorModal/selectors";

interface Props {
  print: boolean;
  errorModal: {
    isOpen: boolean;
    message: string;
  };
}

export const Home: React.FC<Props> = ({ print, errorModal }) => (
  <Fragment>
    <Container>
      <Form />
      {print && <Answer />}
    </Container>
    <ErrorModal show={errorModal.isOpen} text={errorModal.message} />
  </Fragment>
);

const mapStateToProps = (state: StoreState) => ({
  print: getPrint(state),
  errorModal: getErrorModal(state)
});

export default connect(mapStateToProps)(Home);
