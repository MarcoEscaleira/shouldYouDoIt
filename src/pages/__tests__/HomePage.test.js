import React from 'react';
import HomePage from '../HomePage';
import { renderWithRedux } from '../../utils/test-utils';

describe('Home Page test suite', () => {
  it('Should render Home Page correctly', () => {
    const { container } = renderWithRedux(<HomePage />);

    expect(container.firstChild).toMatchSnapshot();
  });

  it('Should render Home Page with an Answer', () => {
    const initialState = {
      print: true,
      data: {
        msg: 'do it',
        img: 'someGif.gif'
      },
      oldTextInput: ''
    };

    const { container } = renderWithRedux(<HomePage />, initialState);

    expect(container.firstChild).toMatchSnapshot();
  });
});
