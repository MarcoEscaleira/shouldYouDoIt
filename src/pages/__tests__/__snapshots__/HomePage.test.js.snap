// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home Page test suite Should render Home Page correctly 1`] = `
<div
  class="container"
>
  <div
    class="questionForm"
  >
    <h1
      class="questionForm__title"
    >
      What do you want to do?
    </h1>
    <form
      autocomplete="off"
      class="questionForm__form"
    >
      <input
        class="questionForm__form__input"
        data-testid="input"
        name="toDo"
        type="text"
        value=""
      />
      <button
        class="questionForm__form__btn"
        type="submit"
      >
        Decide
      </button>
    </form>
  </div>
  <div
    class="modal modal--invisible"
  >
    <div
      class="modal__overlay"
    >
      <div
        class="modal__content"
      >
        <h3
          class="modal__title"
        >
          Error
        </h3>
        <p
          class="modal__body"
        />
        <button
          class="modal__button"
        >
          Okay
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`Home Page test suite Should render Home Page with an Answer 1`] = `
<div
  class="container"
>
  <div
    class="questionForm"
  >
    <h1
      class="questionForm__title"
    >
      What do you want to do?
    </h1>
    <form
      autocomplete="off"
      class="questionForm__form"
    >
      <input
        class="questionForm__form__input"
        data-testid="input"
        name="toDo"
        type="text"
        value=""
      />
      <button
        class="questionForm__form__btn"
        type="submit"
      >
        Decide
      </button>
    </form>
  </div>
  <div
    class="modal modal--invisible"
  >
    <div
      class="modal__overlay"
    >
      <div
        class="modal__content"
      >
        <h3
          class="modal__title"
        >
          Error
        </h3>
        <p
          class="modal__body"
        />
        <button
          class="modal__button"
        >
          Okay
        </button>
      </div>
    </div>
  </div>
  <div
    class="answer"
  >
    <h1
      class="answer__title"
    >
      do it
    </h1>
    <img
      alt="Should You Do It gif"
      class="answer__image"
      src="someGif.gif"
    />
    <button
      class="answer__btn"
    >
      Clean
    </button>
  </div>
</div>
`;
